### ISSUE TITLE
Fatal error: 'ctype.h' file not found

### ISSUE THREAD

**Issue by hugoam** [emscripten-core/emscripten#13109]
Hello !
I'm facing an issue building a c++ file that includes `ctype.h` which so far eludes me.
This is on a fresh ubuntu xenial (WSL on windows) install with almost only emsdk installed (apart from `build-essential`).
I haven't tried building this project since april, so there are probably a bunch of changes in Emscripten that could explain this, but it might also be in my configuration ?

Here is the full output of the compilation that produces the issue, if you spot something that could help explain the cause:
```
In file included from ../../../../bimg/src/image_decode.cpp:20:
/home/hugoam/emsdk/upstream/emscripten/system/include/compat/ctype.h:15:15: fatal error: 'ctype.h' file not found
#include_next <ctype.h>
              ^~~~~~~~~
1 error generated.
em++: error: '/home/hugoam/emsdk/upstream/bin/clang++ -DEMSCRIPTEN -fignore-exceptions -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr -Xclang -isystem/home/hugoam/emsdk/upstream/emscripten/system/include/SDL -target wasm32-unknown-emscripten -D__EMSCRIPTEN_major__=2 -D__EMSCRIPTEN_minor__=0 -D__EMSCRIPTEN_tiny__=11 -D_LIBCPP_ABI_VERSION=2 -Dunix -D__unix -D__unix__ -Werror=implicit-function-declaration --sysroot=/home/hugoam/emsdk/upstream/emscripten/system -Xclang -isystem/home/hugoam/emsdk/upstream/emscripten/cache/wasm/include -Xclang -iwithsysroot/include/libcxx -Xclang -iwithsysroot/lib/libcxxabi/include -Xclang -iwithsysroot/include/compat -Xclang -iwithsysroot/include/libc -Xclang -iwithsysroot/lib/libc/musl/arch/emscripten -Xclang -iwithsysroot/local/include -Xclang -iwithsysroot/include/SSE -Xclang -iwithsysroot/include/neon -Xclang -iwithsysroot/lib/compiler-rt/include -Xclang -iwithsysroot/lib/libunwind/include -MMD -MP -MP -D__STDC_LIMIT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_CONSTANT_MACROS -D_DEBUG -I../../../../bx/include -I../../../../bimg/include -I../../../../bimg/3rdparty -Wall -Wextra -ffast-math -fomit-frame-pointer -g -m32 -std=c++14 -fno-rtti -fno-exceptions -Wunused-value -Wundef -o ../../wasm/obj/x32/Debug/bimg_decode/bimg/src/image_decode.o -c ../../../../bimg/src/image_decode.cpp -c -o ../../wasm/obj/x32/Debug/bimg_decode/bimg/src/image_decode.o' failed
```

The way I understand it, all the files needed by emscripten are actually inside the emscripten folders, including that `ctype.h` file (which I suppose should be the one here: `emsdk/upstream/emscripten/system/include/libcxx/ctype.h` which seems wrapped by the one in the `compat` folder). So I don't really understand what could cause Emscripten to fail to pick it up

**Comment by sbc100** [comment]
Interesting.  Seems to work just for me: 

```
$ cat test.cpp 
#include <ctype.h>
$ em++ -c test.cpp
```

If you run em++ with -v you will see the include path and it should contain 3 different version of ctype.h that use include_next to include one after the other:

```
system/include/libcxx
system/include/compat
system/include/libc
```

For example on my machine `-v` shows:

```
#include "..." search starts here:
#include <...> search starts here:
 /home/sbc/dev/wasm/emscripten/system/include/SDL
 /home/sbc/dev/wasm/emscripten/system/include/libcxx
 /home/sbc/dev/wasm/emscripten/system/lib/libcxxabi/include
 /home/sbc/dev/wasm/emscripten/system/include/compat
 /home/sbc/dev/wasm/emscripten/system/include/libc
 /home/sbc/dev/wasm/emscripten/system/lib/libc/musl/arch/emscripten
 /home/sbc/dev/wasm/emscripten/system/local/include
 /home/sbc/dev/wasm/emscripten/system/include/SSE
 /home/sbc/dev/wasm/emscripten/system/include/neon
 /home/sbc/dev/wasm/emscripten/system/lib/compiler-rt/include
 /home/sbc/dev/wasm/emscripten/system/lib/libunwind/include
 /home/sbc/dev/wasm/emsdk/upstream/lib/clang/12.0.0/include
 /home/sbc/dev/wasm/emscripten/system/include
End of search list.
```

On your machine it looks like the one in the compat directory has been found first for time reason and the libc one is failing to be found afterwards.

I can't see anything obviously wrong.    Can you should the failing em++ command perhaps?

**Comment by tommitytom** [comment]
@hugoam I think this is a bgfx issue as I'm experiencing exactly the same issue, but only when compiling in debug mode, release is fine.

**Comment by bdero** [comment]
Fixed in bkaradzic/bimg#51.

This was due to bimg importing `ctype.h` with the `compat/ctype.h` path directly rather than `ctype.h`. This causes the search cursor to skip past `system/include/compat` and land on `system/include`, which then results in the subsequent `import_next` lookup failing.

My best guess is that a 3rd/redundant path which offered the same contents as `system/include` was removed in an emscripten release late last year, and that importing via `compat/ctype.h` just happened to work unintentionally before that. But I didn't bother bisecting emscripten releases to figure it out because I think importing the header via the `compat/ctype.h` path directly is unequivocally incorrect usage. 🤷‍♂️

Perhaps the compat headers should be moved to a directory adjacent to `system/include`, such as `system/compat/include` (as opposed to `system/include/compat`) to avoid confusion by making it impossible to accidentally import the wrong way.

**Comment by hugoam** [comment]
Great, so it was indeed a bgfx issue and not an emscripten issue, thanks guys and sorry for the noise !
